### PR TITLE
Create Miniconda setup script

### PR DESCRIPTION
**Pull request by kgerheiser**
Setting up Miniconda is kind of tedious, so I created this script to automate the download/install/modulefile as described in the [instructions](https://github.com/NOAA-EMC/spack-stack/blob/develop/doc/source/Prerequisites.rst).

Now you just have to run `./bootstrap.sh -p <dir>` and the script will create a standard directory structure, then download Miniconda, install it along with Poetry, etc, and then create a modulefile.

### CODE REVIEW

File: bootstrap/bootstrap.sh
Comment (climbfuji) [inline bootstrap/bootstrap.sh:14]: Should we follow the conventions and use `--prefix` instead/in addition?

File: bootstrap/bootstrap.sh
Comment (climbfuji) [inline bootstrap/bootstrap.sh:21]: Good to pin the version! Is that `python-3.9.7`?

**Review by climbfuji — COMMENTED** [review]
I like that!

What is the intended directory structure for this setup?

If I read the script correctly, it not only sets up the directory tree for miniconda, but also imposes where envs, mirrors, ... go. Where would you check out spack-stack? In `src`?

File: bootstrap/bootstrap.sh
Comment (kgerheiser) [reply bootstrap/bootstrap.sh:14]: Sure

File: bootstrap/bootstrap.sh
Comment (kgerheiser) [reply bootstrap/bootstrap.sh:21]: It's actually 3.9.12. They don't encode the minor version.

**Comment by kgerheiser** [comment]
> What is the intended directory structure for this setup?

I'm getting a little ahead of myself with these extra directories like mirrors.

I envision a structure like:

```
<prefix>
  /mirrors
  /envs
  /apps
  /modulefiles
  /src
```

And yeah I was thinking `spack-stack` would go in src or maybe `src/spack-stack/<env-name>`  because you may have multiple sources.

File: bootstrap/bootstrap.sh
Comment (kgerheiser) [reply bootstrap/bootstrap.sh:14]: getopts does not support long args. I would just leave it as `-p`. There's only a single argument.

Comment (climbfuji) [reply bootstrap/bootstrap.sh:14]: Ok

File: bootstrap/bootstrap.sh
Comment (climbfuji) [reply bootstrap/bootstrap.sh:21]: Thanks!

File: bootstrap/templates/miniconda
Comment (climbfuji) [inline bootstrap/templates/miniconda:22]: We'll need to take care to update this when we update the Python version. Maybe put a comment above the line where you specify the miniconda version you want to use?

**Review by climbfuji — COMMENTED** [review]
I am happy with this script, makes life a lot easier. But I would prefer to agree on (and write down somewhere) the target directory structure before we merge this changes, if that's ok.

**Comment by kgerheiser** [comment]
Updated the documentation with a directory structure section

**Comment by climbfuji** [comment]
> Updated the documentation with a directory structure section

I was going to review that, but I can't see the doc update, unfortunately.

**Comment by kgerheiser** [comment]
Fixed, changes weren't committed.

File: bootstrap/templates/git-lfs
Comment (climbfuji) [inline bootstrap/templates/git-lfs]: If you are moving the git-lfs template, then you also need to update the documentation. I recommend moving only the miniconda template for now. In the original location doc/modulefile_templates, there are now 5 templates altogether (ecflow, git-lfs, miniconda, openmpi, qt). I would only move the miniconda template and check if there is any documentation that needs to be updated.

File: doc/source/MaintainersSection.rst
Comment (climbfuji) [inline doc/source/MaintainersSection.rst]:
```suggestion
* modulefiles - Prerequisite modules not built with Spack.
```
```suggestion
* modulefiles - Prerquisite modules not built with Spack.
```

File: bootstrap/templates/git-lfs
Comment (kgerheiser) [reply bootstrap/templates/git-lfs]: I had added the ability to download git-lfs as well, but dropped it and then these new modules were added in the meantime. I removed git-lfs from this directory.

**Review by climbfuji — COMMENTED** [review]
I am a little worried about the directory structure. If we want to be able to install major spack-stack versions in separate directories, then it will be difficult to identify which environment in the `envs` directory corresponds to which `spack-stack` version in the `src` directory.

If, on the other hand, we decided to make the top-level directory (where `apps` etc resides in) dependent on the spack-stack version, then we end up with a lot of duplicates for the apps/modulefiles that are generated outside of spack-stack, then we end up with a lot of redundancy.

**Comment by kgerheiser** [comment]
I have had the same thoughts about grouping environments and having multiple spack-stack sources. What if we add another directory in `envs` that corresponds to the spack-stack version?

```
envs/
   v1.0.0/
      skylab-1.0.0
      another-env
   v2.0.0
       skylab-2.0.0
...
src/
   spack-stack-v1.0.0
   spack-stack-v2.0.0
```

Then the src and envs dir correspond.

**Comment by kgerheiser** [comment]
Or 

```
envs/
   v1.0.0/
      src/spack-stack
      skylab-1.0.0
      another-env
   v2.0.0
       src/spack-stack
       skylab-2.0.0
```

Then you can have multiple environments with a single spack-stack source

**Comment by climbfuji** [comment]
> I have had the same thoughts about grouping environments and having multiple spack-stack sources. What if we add another directory in `envs` that corresponds to the spack-stack version?
> 
> ```
> envs/
>    v1.0.0/
>       skylab-1.0.0
>       another-env
>    v2.0.0
>        skylab-2.0.0
> ...
> src/
>    spack-stack-v1.0.0
>    spack-stack-v2.0.0
> ```
> 
> Then the src and envs dir correspond.

Yes, I had similar thoughts. How do we handle minor versions that get installed into the existing environments (or do we not want to do that anymore)? Should those directories in `envs` just be called `spack-stack-v1`/`spack-stack-1` and everything that is `spack-stack-1.x.y` gets installed underneath there?

Or be super pragmatic and say that every new version, whether it is major/minor/patch, requires a full new build of every environment?

**Comment by kgerheiser** [comment]
> Should those directories in envs just be called spack-stack-v1/spack-stack-1 and everything that is spack-stack-1.x.y gets installed underneath there?

I like that

> Or be super pragmatic and say that every new version, whether it is major/minor/patch, requires a full new build of every environment?

We could make use of chained installations for minor updates. A v1.1.0 uses v1.0.0 as an upstream. Then, we don't have to wait hours for a minor update and we don't have to do a full re-install until we get to v2.0.0 (which let's say has a new NetCDF)

**Comment by climbfuji** [comment]
Problem with the chained installs it that it doesn't create the modules for the upstream repos, means that users have to use two `module use` commands, and it doesn't integrate with our meta modules either.

**Comment by kgerheiser** [comment]
I like your idea of rolling releases every x months or so. I say we can append versions to an existing install during that stack's lifetime.

Then, in the next stack we try to include any one-off installs as the new base version, if possible. Like someone says we need ESMF 8.4.0, we install that alongside 8.3.0 in spack-stack 1.1.0, and then when the next release comes around we install 8.4.0 as the default and drop 8.3.0.

And this information should be encapsulated in a `spack.yaml`.

**Comment by climbfuji** [comment]
@kgerheiser So is this what we want?
```
envs/
   v1/
      skylab-1.0.0
      another-env
   v2/
      skylab-2.0.0
...
src/
   spack-stack-v1.0.0
   spack-stack-v1.0.1 # updates envs/v1/...
   spack-stack-v1.1.0 # updates envs/v1/...
   ...
   spack-stack-v2.0.0
   spack-stack-v2.1.0 # updates envs/v2/...
```

**Comment by kgerheiser** [comment]
How about this @climbfuji?

**Comment by kgerheiser** [comment]
@climbfuji could you review my new changes? I updated the docs based on our previous conversation and added an example.

**Review by climbfuji — APPROVED** [review]
I think that's great and we should start working with this from the next major spack-stack release on (v2?) to not break the current directory structure for 1.0.0/1.0.1 (?).

**Comment by kgerheiser** [comment]
I can install it like this on Hera/Jet starting now, right?  The directory structure is an implementation detail for users.

**Comment by climbfuji** [comment]
> I can install it like this on Hera/Jet starting now, right? The directory structure is an implementation detail for users.

I was hoping to do that from version 2 on, so that the version 1 installs are identical everywhere.

**Comment by climbfuji** [comment]
> > I can install it like this on Hera/Jet starting now, right? The directory structure is an implementation detail for users.
> 
> I was hoping to do that from version 2 on, so that the version 1 installs are identical everywhere.

But it's ok if Discover and Orion are different for v1

**Comment by kgerheiser** [comment]
I can install the skylab-1.0.0 release using the same layout for consistency

**Comment by climbfuji** [comment]
> I can install the skylab-1.0.0 release using the same layout for consistency

Thanks, Kyle. I'll do the same on Cheyenne.

**Pull request merged**
